suite/tests/linux/mangle_asynch.c

```diff
@@ -0,0 +1,173 @@
+#    include <setjmp.h>
+#    include <signal.h>
+#    include <pthread.h>
+
+volatile bool test_ready = false;
+volatile bool test_done = false;
+volatile int loop_inc = 1;
+
+/* asm routines */
+void
+test_asm();
+
+static SIGJMP_BUF mark;
+
+static void
+handle_signal(int signal, siginfo_t *siginfo, ucontext_t *ucxt)
+{
+#    ifdef X86_64
+    if (signal == SIGILL) {
+        sigcontext_t *sc = SIGCXT_FROM_UCXT(ucxt);
+        if (sc->LOOP_COUNT_REG_SIG != LOOP_COUNT)
+            print("ERROR: incorrect result!\n");
+        SIGLONGJMP(mark, 1);
+    } else if (signal != SIGUSR2) {
+        print("Unexpected signal!\n");
+    }
+#    endif
+}
+
+static void *
+thread_routine(void *arg)
+{
+#    ifdef X86_64
+    pthread_t main_thread = *(pthread_t *)arg;
+    while (!test_ready) {
+        /* Empty. */
+    }
+    struct timespec sleeptime;
+    sleeptime.tv_sec = 0;
+    sleeptime.tv_nsec = 1000 * 1111;
+    while (!test_done) {
+        pthread_kill(main_thread, SIGUSR2);
+        nanosleep(&sleeptime, NULL);
+    }
+#    endif
+    return NULL;
+}
+
+int
+main(int argc, const char *argv[])
+{
+#    define NUM_THREADS 8
+    pthread_t thread[NUM_THREADS];
+    void *retval;
+    pthread_t main_thread = pthread_self();
+
+    for (int i = 0; i < NUM_THREADS; ++i) {
+        if (pthread_create(&thread[i], NULL, thread_routine, &main_thread) != 0) {
+            perror("Failed to create thread");
+            exit(1);
+        }
+    }
+
+    intercept_signal(SIGILL, (handler_3_t)&handle_signal, false);
+    intercept_signal(SIGUSR2, (handler_3_t)&handle_signal, false);
+
+    /* Test loop counter using mangled rip-rel instruction, interrupted by asynch
+     * signals.
+     */
+    if (SIGSETJMP(mark) == 0) {
+        test_asm();
+    }
+
+    for (int i = 0; i < NUM_THREADS; ++i) {
+        if (pthread_join(thread[i], &retval) != 0)
+            perror("Failed to join thread");
+    }
+
+    print("Test finished\n");
+    return 0;
+}
+
+#else /* asm code *************************************************************/
+#    include "asm_defines.asm"
+#    include "mangle_asynch-shared.h"
+/* clang-format off */
+START_FILE
+
+#ifdef X64
+# define FRAME_PADDING 0
+#else
+# define FRAME_PADDING 0
+#endif
+
+#define FUNCNAME test_asm
+        DECLARE_FUNC_SEH(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+#ifdef X86
+#ifdef X64
+        PUSH_CALLEE_SAVED_REGS()
+        sub      REG_XSP, FRAME_PADDING /* align */
+        END_PROLOG
+
+        jmp      test
+     test:
+        mov      BYTE SYMREF(test_ready), HEX(1)
+        mov      LOOP_TEST_REG_ASM, LOOP_COUNT
+        mov      LOOP_COUNT_REG_ASM, 0
+      loop:
+        /* The address will get mangled into REG_XAX. */
+        add      LOOP_COUNT_REG_ASM, PTRSZ SYMREF(loop_inc)
+        sub      LOOP_TEST_REG_ASM, 1
+        cmp      LOOP_TEST_REG_ASM, 0
+        jnz      loop
+        jmp      epilog
+     epilog:
+        mov      BYTE SYMREF(test_done), HEX(1)
+        ud2
```

Comment:
Comments would help to understand this code: this will trigger the SIGILL handler which ensures there was no extra add within the loop from re-execution due to translating back.

Reply:
Done